src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+  /**
+    * Connects to Zookeeper and offers leadership; monitors leader state. Watches for leadership changes (leader
+    * changed, was elected leader, lost leadership), and emits events accordingly.
+    *
+    * Materialized cancellable is used to abdicate leadership; which will do so followed by a closing of the stream.
+    */
+  def apply(
+    clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+    zooKeeperLeaderPath: String,
+    zooKeeperConnectionTimeout: FiniteDuration,
+    hostPort: String,
+    singleThreadExecutor: ExecutorService): Source[LeadershipState, Cancellable] = {
+    val singleThreadEC = ExecutionContext.fromExecutor(singleThreadExecutor)
+    Source.queue[LeadershipState](16, OverflowStrategy.dropHead)
+      .mapMaterializedValue { sq =>
+        val emitterLogic = new CuratorEventEmitter(singleThreadExecutor, clientCloseable, zooKeeperLeaderPath, hostPort, sq)
+        emitterLogic.start()
+        sq.watchCompletion().onComplete { _ => emitterLogic.cancel() }(singleThreadEC)
+        emitterLogic
+      }
+      .initialTimeout(zooKeeperConnectionTimeout)
+      .concat(Source.single(LeadershipState.Standby(None)))
+      .via(EnrichedFlow.dedup(LeadershipState.Standby(None)))
+      .map { e =>
+        e match {
+          case LeadershipState.ElectedAsLeader =>
+            logger.info("Leader won.")
+          case LeadershipState.Standby(currentLeader) =>
+            logger.info(s"Leader defeated. Current leader: ${currentLeader.getOrElse("-")}")
+        }
+        e
+      }
+  }
+
+  @SuppressWarnings(Array("CatchThrowable"))
+  private class CuratorEventEmitter(
+      singleThreadExecutor: ExecutorService,
+      clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+      zooKeeperLeaderPath: String,
+      hostPort: String,
+      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+
+    val client = clientCloseable.closeable
+    private lazy val singleThreadEC: ExecutionContext = ExecutionContext.fromExecutor(singleThreadExecutor)
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    @volatile private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* It is important that we re-register our watch _BEFORE_ we get the current the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = Future {
```

Comment:
Return type should be `Future[Done?]`

Reply:
No. :) Switched to `singleThreadEC.execute { () => ` for clarity.

With Scala 2.11 we had to instantiate the Runnable manually but with 2.12 we can just pass a Function0 🎉 🎉 🎉